### PR TITLE
Support refactorings through code actions.

### PR DESCRIPTION
**Pull request by doriath**
- make LspCodeAction work on a selected range, to support extract method
or extract variable refactorings
- properly execute commands returned by code actions
- support workspace/applyEdit requests from server

### CODE REVIEW

**Comment by doriath** [comment]
Here is a demo of code actions:

code-actions

File: autoload/lsp/client.vim [hunk 1]
```diff
@@ -82,7 +82,14 @@ function! s:on_stdout(id, data, event) abort
         if exists('l:response')
             " call appropriate callbacks
             let l:on_notification_data = { 'response': l:response }
-            if has_key(l:response, 'id')
+            if has_key(l:response, 'method') && has_key(l:response, 'id')
+                " it is a request from a server
+                let l:request = l:response
+                if l:request['method'] ==# 'workspace/applyEdit'
+                    call lsp#ui#vim#apply_workspace_edits(l:request['params']['edit'])
```
Comment (prabirshrestha) [inline autoload/lsp/client.vim]: we shouldn't call this directly here. can we instead expose a method similar to `on_notification` of `lsp#client#start`.

https://github.com/prabirshrestha/vim-lsp/blob/ee1dd1638a69930fb49be7f7e4e17c9bdb16de63/autoload/lsp.vim#L353

Comment (doriath) [reply autoload/lsp/client.vim]: Extracted to separate `on_request` callback similar to `on_notification`.

File: autoload/lsp/ui/vim.vim [hunk 2]
```diff
@@ -505,14 +538,46 @@ function! s:handle_code_action(server, last_req_id, type, data) abort
...
+    call lsp#send_request(a:server, {
+        \ 'method': 'workspace/executeCommand',
+        \ 'params': l:params,
+        \ })
+endfunction
+
 " @params
 "   workspace_edits - https://microsoft.github.io/language-server-protocol/specification#workspaceedit
-function! s:apply_workspace_edits(workspace_edits) abort
+function! lsp#ui#vim#apply_workspace_edits(workspace_edits) abort
```
Comment (prabirshrestha) [inline autoload/lsp/ui/vim.vim]: Can we move this function to `lsp#ui#vim#workspace_edits#apply_workspace_edits` similar to how we have `lsp#utils#text_edit#apply_text_edits`

Comment (doriath) [reply autoload/lsp/ui/vim.vim]: I moved it to `lsp#utils#workspace_edits#apply_workspace_edit` to match `text_edit`. Please let me know if you would prefer `lsp#ui#vim` over `lsp#utils`.

Comment (prabirshrestha) [reply autoload/lsp/ui/vim.vim]: `lsp#utils` is better.

**Comment by prabirshrestha** [comment]
Merged. This is a great improvement. Thanks.

**Pull request merged**